### PR TITLE
V15: Hybrid Caching

### PR DESCRIPTION
**Pull request by Zeegaan**
# Hybrid Caching
This PR introduces new hybrid cache implementations of the cache interfaces,
If you want to learn more about what hybrid caching is, read: https://learn.microsoft.com/en-us/aspnet/core/performance/caching/hybrid?view=aspnetcore-9.0
## Second level cache
For how to implement a secondary cache with this, see: https://learn.microsoft.com/en-us/aspnet/core/performance/caching/distributed?view=aspnetcore-9.0#distributed-redis-cache

## Additions to the cache interfaces
- We have the original cache interfaces in the Core project:
- `IPublishedMemberCache`
- `IPublishedMediaCache`
- `IPublishedContentCache`

And with this PR, we add some async methods to these. This way we break less code, but we can also modernize with some async methods.

## What happened to PublishedSnapshot?
With nucache, we had a concept of a `PublishedSnapshot`, this was a cache snapshot, so that you would have the same entities during a request, so that if anything got changed /deleted during your request, that would not interfere / cause errors.
This is no longer a thing with the Hybrid Cache, as it has no concept of cache snapshots.
This also means that `IPublishedSnapshot`, `IPublishedSnapshotAccessor` & `SnapshotCache` is all obsolete.

And even the `PropertyCacheLevel.Snapshot`  is also obsolete.

### Cache implementations
 The cache implementations instead of having tons of logic inside them, now calls the cache service instead to get out items. 

Note that theres a lot of not implemented methods here, as they are all obsolete, and no longer part of this caching. 
Thus they will have to be removed when we remove NuCache.

What Caches do we have now then?

- Content Cache
- Media Cache
- Member Cache
- Elements Cache - This is used for all properties that have `PropertyCacheLevel.Elements`
- Domain Cache - Caching of domains

### The cache services
These cache services that the caches use, now use the new snazzy dotnet 9 feature `HybridCache`.

Cache service responsibility includes:

- Cache responsibilities such as:
    - Getting entities from the hybrid cache ( if not in the cache, use the nucache content repository to fetch from database)
    - Refreshing a node in the cache (aka just deleting it when we know a node has updated)
    - Seeding the database with a given list of contenttype ids
- Mapping the content cache node you get from the cache, over to `IPublishedContent`
- Mapping keys to Ids and vice versa. We need both of them, as the given key is used to form the cache key, and the id is used to get the entity from the database.

### PublishedContent changes
PublishedContent no longer has any Tree logic, this means, we no longer have children / parent.

### PublishedProperty
The big change to PublishedProperty is that we no longer use snapshots, so in the `GetCacheValues` method, we can avoid using snapshot, and instead just fall back to the `Element` behavior.

## Seeding
So for NuCache, it would load everything into memory, this is really performant when everything is finally loaded in. But causes  issues like :
- Long boot times on larger sites
- Memory issues as it would take up too much memory
- Corrupted cache

So we have strayed from this approach with this. We do however recognize that sometimes you might want to have some pages always in memory (stuff like the front page), this is where the seeding feature comes in.

You can now instead deine a list of content type ids in `appsettings.json`, which in turn will load every document that uses the given document types, into the cache, and will never evict them.

This new setting can be set like:

```
 "Umbraco": {
    "CMS": {
      "Cache": {
         "ContentTypeIds": [1068, 1069]
       },
```

# Features still missing
- When updating a seeded entity, it is no longer permanently in the cache, we need to figure out a way to re-get these on deletion and have the entry options set to a year again.
- 
# How to test
Note: We can't do a full proper tests before we can replace NuCache. And we can only replace NuCache once this and https://github.com/umbraco/Umbraco-CMS/pull/16818 is merged 😁 
For now we can instead test this with trying out some distributed cache implementation, and asserting entities actually are in the cache, both after getting the entity, but also seeding.

- Create a document type named "RootSeed" (note down the integer id of this content type), with allow as root set to true.
- Create another document type "Root" with allow as root set to true
- Create one piece of content with each. please note down the id of the content you create with "Root"
- Stop the site
- Go to your appsettings, and put in the cache setting to enable seeding:

```
 "Umbraco": {
    "CMS": {
      "Cache": {
         "ContentTypeKeys": [Put Key of "RootSeed" here instead.]
       },
```
We will now use the SqlServer distributed cache, you can directly follow the documentation here: https://learn.microsoft.com/en-us/aspnet/core/performance/caching/distributed?view=aspnetcore-8.0#distributed-sql-server-cache
It would also be nice to tests this with other implementations of this, like Redis.
- First we will have to create the SQL database, i used `(localdb)\MSSQLLocalDB` and created a database called `DistCache`
- Creating the cache with this command: `dotnet sql-cache create "Data Source=(localdb)\MSSQLLocalDB;Initial Catalog=DistCache;Integrated Security=True;" dbo TestCache`
- Now install the NuGet package: `Microsoft.Extensions.Caching.SqlServer` to the Web.UI project
- Now that the nuget package is installed, we need to register the Distributed cache in `program.cs`, paste this on line 2:
```
builder.Services.AddDistributedSqlServerCache(options =>
{
    options.ConnectionString = "Data Source=(localdb)\\MSSQLLocalDB;Database=DistCache;Integrated Security=True;";
    options.SchemaName = "dbo";
    options.TableName = "TestCache";
});
```

- Lastly to trigger the caching of the node you created with "root", we need a controller endpoint we can call (this could also be a notificationhandler or wherever you want to trigger this code)

``` 
using Microsoft.AspNetCore.Mvc;
using Umbraco.Cms.Core.Models.PublishedContent;
using Umbraco.Cms.Core.PublishedCache;

namespace Umbraco.Cms.Web.UI;

[ApiController]
[Route("[controller]")]
public class MyController : ControllerBase
{
    private readonly IPublishedContentCache _publishedContentCache;

    public MyController(IPublishedContentCache publishedContentCache)
    {
        _publishedContentCache = publishedContentCache;
    }

    [HttpGet]
    public ActionResult Get()
    {
        IPublishedContent? content = _publishedContentCache.GetById("PUT ID OF CONTENT CREATED WITH ROOT HERE");
        return Ok();
    }
}
```

- Now that the setup is done, run your umbraco again
- After umbraco has started up, go to the database and look at the `dbo.TestCache` table, this should now contain 1 entry, and that should be your content you created with the "RootSeed" contentype,
- If you then call your controller (using postman, send a request to `https://localhost:44339/My`)
- You should now see 2 entries in your database table.

### CODE REVIEW

File: Umbraco.PublishedCache.HybridCache/CacheManager.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+    public IPublishedContentCache? Content { get; }
+
+    public IPublishedMediaCache? Media { get; }
+
+    public IPublishedMemberCache? Members { get; }
+
+    public IDomainCache? Domains { get; }
+
+    public IAppCache? ElementsCache { get; }
```
[the remark this answers is not in the document]
Comment (Zeegaan) [reply Umbraco.PublishedCache.HybridCache/CacheManager.cs:26]: They shouldn't be, good catch 👁️ 👃 👁️

File: Umbraco.PublishedCache.HybridCache/ContentNode.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    public readonly int Id;
+
+    private ContentData? _draftData;
+
+    // draft and published version (either can be null, but not both)
+    // are models not direct PublishedContent instances
+    private IPublishedContent? _draftModel;
+    private ContentData? _publishedData;
+    private IPublishedContent? _publishedModel;
+    private IPublishedSnapshotAccessor? _publishedSnapshotAccessor;
```
[the remark this answers is not in the document]
Comment (Zeegaan) [reply Umbraco.PublishedCache.HybridCache/ContentNode.cs:22]: They are gone  ✂️

File: Umbraco.PublishedCache.HybridCache/Factories/CacheNodeFactory.cs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            }
+        }
+
+        return new ContentData(
+            content.Name,
+            null,
+            content.VersionId,
+            content.UpdateDate,
+            content.CreatorId,
+            -1,
```
[the remark this answers is not in the document]
Comment (Zeegaan) [reply Umbraco.PublishedCache.HybridCache/Factories/CacheNodeFactory.cs]: This was just a bug, templates was never set correctly... fixed! 🚀

File: Umbraco.PublishedCache.HybridCache/Factories/PublishedContentFactory.cs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    public IPublishedMember ToPublishedMember(IMember member)
+    {
+        IPublishedContentType contentType = _contentTypeCache.Get(PublishedItemType.Member, member.ContentTypeId);
+        var contentData = new ContentData(
+            member.Name,
+            null,
+            0,
+            member.UpdateDate,
+            member.CreatorId,
+            -1,
```
[the remark this answers is not in the document]
Comment (Zeegaan) [reply Umbraco.PublishedCache.HybridCache/Factories/PublishedContentFactory.cs]: Added clarifying comment 😁

File: Umbraco.PublishedCache.HybridCache/Factories/PublishedContentFactory.cs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    private Dictionary<string, PropertyData[]> GetPropertyValues(IPublishedContentType contentType, IMember member)
+    {
+        var properties = member
+            .Properties
+            .ToDictionary(
+                x => x.Alias,
+                x => new[] { new PropertyData { Value = x.GetValue(), Culture = string.Empty, Segment = string.Empty } },
+                StringComparer.OrdinalIgnoreCase);
+
+        // see also PublishedContentType
```
[the remark this answers is not in the document]
Comment (Zeegaan) [reply Umbraco.PublishedCache.HybridCache/Factories/PublishedContentFactory.cs:92]: Thats a wierd comment, changed 😁

File: Umbraco.PublishedCache.HybridCache/ContentCache.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+public sealed class ContentCache : IPublishedContentCache
+{
+    private readonly IContentCacheService _contentCacheService;
+    private readonly PublishedContentTypeCache _contentTypeCache;
+
+    public ContentCache(IContentCacheService contentCacheService, IPublishedContentTypeCacheAccessor publishedContentTypeCacheAccessor)
+    {
+        _contentCacheService = contentCacheService;
+        _contentTypeCache = publishedContentTypeCacheAccessor.Get();
```
[the remark this answers is not in the document]
Comment (Zeegaan) [reply Umbraco.PublishedCache.HybridCache/ContentCache.cs]: ContentTypeCache is now injectibe and the accessor is deleted 👍

File: Umbraco.PublishedCache.HybridCache/Persistence/DatabaseCacheRepository.cs [hunk 7]
```diff
@@ -184,14 +184,14 @@ AND cmsContentNu.nodeId IS NULL
         return count == 0;
     }
 
-    public ContentCacheNode? GetContentSource(int id, bool preview = false)
+    public async Task<ContentCacheNode?> GetContentSource(int id, bool preview = false)
```
Comment (nikolajlauridsen) [inline Umbraco.PublishedCache.HybridCache/Persistence/DatabaseCacheRepository.cs:187]: We should rename these to end in async as well 😄

File: src/Umbraco.Cms.Api.Management/Umbraco.Cms.Api.Management.csproj [hunk 8]
```diff
@@ -14,6 +14,7 @@
   </ItemGroup>
 
   <ItemGroup>
+    <ProjectReference Include="..\..\Umbraco.PublishedCache.HybridCache\Umbraco.PublishedCache.HybridCache.csproj" />
```
Comment (nikolajlauridsen) [inline src/Umbraco.Cms.Api.Management/Umbraco.Cms.Api.Management.csproj:17]: The project is in the wrong folder, it should be in `src` but it's in the `Umbraco-CMS` folder 🙈

File: src/Umbraco.Core/Models/CacheSettings.cs [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace Umbraco.Cms.Core.Models;
+
+[UmbracoOptions(Constants.Configuration.ConfigCache)]
+public class CacheSettings
+{
+    /// <summary>
+    ///     Gets or sets a value for the collection of content type ids to always have in the cache.
+    /// </summary>
+    public List<int> ContentTypeIds { get; set; } =
```
Comment (nikolajlauridsen) [inline src/Umbraco.Core/Models/CacheSettings.cs:11]: We should consider using keys instead here.

File: src/Umbraco.Cms.Api.Management/Umbraco.Cms.Api.Management.csproj (hunk 8 above)
Comment (Zeegaan) [reply src/Umbraco.Cms.Api.Management/Umbraco.Cms.Api.Management.csproj]: What the... 🙈 
Moved 🚀

File: src/Umbraco.Core/Models/CacheSettings.cs (hunk 9 above)
Comment (Zeegaan) [reply src/Umbraco.Core/Models/CacheSettings.cs]: It is now indeed using keys 🚀

File: src/Umbraco.PublishedCache.HybridCache/Persistence/ContentSourceDto.cs [hunk 10]
```diff
@@ -0,0 +1,67 @@
+﻿using Umbraco.Cms.Core.Models;
+
+namespace Umbraco.Cms.Infrastructure.HybridCache.Persistence
+{
+    // read-only dto
+    internal class ContentSourceDto : IReadOnlyContentBase
+    {
+        public int Id { get; set; }
```
Comment (nikcio) [inline src/Umbraco.PublishedCache.HybridCache/Persistence/ContentSourceDto.cs:8]: Hey if this is really read only shouldn't the setters then be removed?

For example with a constructor:
```cs
public int Id { get; }
```

or

```cs
public required int Id { get; init; }
```

If it has to be set on creation. (The `required` keyword is optional and just means it has to be set to a value)

I would also imagine that you would want this for the classes that uses `[ImmutableObject(true)]`?

Just some a thought I had when looking it over 😄

Comment (nikolajlauridsen) [reply src/Umbraco.PublishedCache.HybridCache/Persistence/ContentSourceDto.cs:8]: That's a good point, I've updated it now 👍 Thanks

**Comment by nikolajlauridsen** [comment]
No, there has been no change in this regard

**Pull request merged**